### PR TITLE
feat(frost): Add function for round3

### PR DESCRIPTION
**Pull request by rohanjadvani**

### CODE REVIEW

File: src/dkg/round3.rs [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    let expected_round1_checksum = round1::input_checksum(
+        min_signers,
+        round1_public_packages.iter().map(|pkg| pkg.identity()),
+    );
+
+    let mut identities = BTreeMap::new();
+    let mut round1_frost_packages: BTreeMap<Identifier, Round1Package> = BTreeMap::new();
+    for public_package in round1_public_packages.clone() {
```
Comment (hughy) [inline src/dkg/round3.rs:59]: should we move some of these preprocessing steps into util functions that we can use in both round2 and round3?

File: src/dkg/round3.rs [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+use super::error::Error;
+use super::round1;
+use super::round2;
+
+pub fn round3<'a, P, Q>(
+    identity: &Identity,
+    round2_secret_package: &Round2SecretPackage,
+    round1_public_packages: P,
+    round2_public_packages: Q,
+) -> Result<(KeyPackage, PublicKeyPackage), Error>
```
Comment (hughy) [inline src/dkg/round3.rs]: i think we also want to construct and return the GroupSecretKey here

File: src/dkg/round3.rs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    let round2_public_packages = round2_public_packages.into_iter().collect::<Vec<_>>();
+    let expected_round2_checksum =
+        round2::input_checksum(round1_public_packages.iter().map(Borrow::borrow));
+
+    let mut round2_frost_packages: BTreeMap<Identifier, Round2Package> = BTreeMap::new();
+    for public_package in round2_public_packages {
+        if public_package.checksum() != expected_round2_checksum {
+            return Err(Error::ChecksumError(ChecksumError::DkgPublicPackageError));
+        }
+
```
Comment (hughy) [inline src/dkg/round3.rs:107]: do you think it's worthwhile to check here that `self_identity == public_package.recipient_identity`?

File: src/dkg/round3.rs [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+            round1_public_packages.len()
+        )));
+    }
+
+    let expected_round1_checksum = round1::input_checksum(
+        min_signers,
+        round1_public_packages.iter().map(|pkg| pkg.identity()),
+    );
+
+    let mut identities = BTreeMap::new();
```
Comment (hughy) [inline src/dkg/round3.rs]: do we need the `identities` map for round3? it doesn't look like we read from it later on

File: src/dkg/round3.rs (hunk 1 above)
Comment (rohanjadvani) [reply src/dkg/round3.rs:59]: Discussed in huddle - will leave as is for now

File: src/dkg/round3.rs (hunk 3 above)
Comment (rohanjadvani) [reply src/dkg/round3.rs:107]: Added

**Review by hughy — APPROVED** [review]

**Pull request merged**